Include/Acidanthera/Library/OcMiscLib.h

```diff
@@ -263,4 +263,332 @@ OcConsoleFlush (
+  the first element.
+
+  @param[in]      String1       String 1
+  @param[in]      String2       String 2
+
+  @retval Comparison result.
+**/
+INTN
+EFIAPI
+ReverseStrCmp (
+  IN CONST CHAR16 *String1,
+  IN CONST CHAR16 *String2
+  );
+
+/**
+  Sort by ASCII string descending. Use AsciiStrCmp for ascending.
+  Suitable for sorting larger structures, as long as the string pointer is
+  the first element.
+
+  @param[in]      String1       String 1
+  @param[in]      String2       String 2
+
+  @retval Comparison result.
+**/
+INTN
+EFIAPI
+ReverseAsciiStrCmp (
+  IN CONST CHAR8 *String1,
+  IN CONST CHAR8 *String2
+  );
+
+/**
+  Sort array of items by first pointer with comparison function.
+  Currently only suitable for small lists, tens of items.
+
+  @param[in,out]  Items         Array of items to sort.
+  @param[in]      ItemCount     Number of items in the array.
+  @param[in]      ItemSize      Size of each item in the array.
+  @param[in]      CompareItems  Comparison function. The derefenced first pointer
+                                from each item to compare is passed in. (E.g. if the
+                                first element in each item is CHAR16 * then StrCmp
+                                would be a suitable sort function.)
+
+  @retval EFI_SUCCESS           List was sorted.
+  @retval EFI_OUT_OF_RESOURCES  Out of memory.
+**/
+EFI_STATUS
+OcSortByFirstPointer (
+  IN OUT        VOID          *Items,
+  IN      CONST UINTN         ItemCount,
+  IN      CONST UINTN         ItemSize,
+  IN      CONST OC_COMPARE    CompareItems
+  );
+
+/**
+  Sort array of items by comparison function.
+  Currently only suitable for small lists, tens of items.
+
+  @param[in,out]  Items         Array of items to sort.
+  @param[in]      ItemCount     Number of items in the array.
+  @param[in]      ItemSize      Size of each item in the array.
+  @param[in]      CompareItems  Comparison function. The pointer to each item
+                                itself is passed in.
+
+  @retval EFI_SUCCESS           List was sorted.
+  @retval EFI_OUT_OF_RESOURCES  Out of memory.
+**/
+EFI_STATUS
+OcSort (
+  IN OUT        VOID          *Items,
+  IN      CONST UINTN         ItemCount,
+  IN      CONST UINTN         ItemSize,
+  IN      CONST OC_COMPARE    CompareItems
+  );
+
+/*
+  Forward declaration of OC_DYNAMIC_ARRAY structure.
+*/
+typedef struct OC_DYNAMIC_ARRAY_ OC_DYNAMIC_ARRAY;
+
+/**
+  Free any allocated memory pointed to by dynamic array item.
+
+  @param[in]      Item          Item to free.
+**/
+typedef
+VOID
+(* OC_DYNAMIC_ARRAY_FREE_ITEM) (
+  IN VOID *Item
+  );
+
+/**
+  Utility method to free memory pointed to by dynamic array
+  item when the item is a single pointer (or when the only
+  allocated memory is pointed to by a pointer which is the
+  first element).
+
+  @param[in]      Item          Dynamic array item to free.
+**/
+VOID
+OcDynamicArrayFreePointerItem (
```

Comment:
Not entirely sure if it is worth using in OcXmlLib, since it adds 8 bytes overhead per XML node, as well as (I suppose) more work and the risk of introducing bugs unnecessarily for possibly not enough gain?

Perhaps could just add comment there that these might be combined?

Reply:
Naming: OcVectorLib sounds nice, but I believe misleadingly implies that it has something to do with vector operations (dot product, cross product, SIMD, interaction with matrices, or some such).

Other possible names:
 - DynamicArray
 - DynamicBuffer
 - ?

Reply:
I meant that OcXmlLib could use this one, not vice versa. OcDynamicArraryLib feels ok, can be also OcPoolArrayLib, OcDynArrayLib, or OcFlexArrayLib. I think I like the last one.

Reply:
💪 lol!

I still wasn't sure whether it was worth digging into and changing insides of OcXmlLib, for reasons stated, but can do...

Reply:
Have added to to-do list in main PR post - could be moved into an issue and done after merge?